### ISSUE TITLE
Create endpoint for frontend to authorize access to their files in the LMS app

### ISSUE THREAD

**Issue by robertknight** [hypothesis/lms#501]
This is a follow-up from https://github.com/hypothesis/lms/issues/470.

Create a backend endpoint (provisionally `/authorize_lms_file_access`) which the frontend will open in a popup window in order to prompt the user to grant the LMS app access to their files in the LMS app's storage. When the authorization flow completes, whether successful or not, the popup window should close itself.

See dummy backend implementation in the lms-file-picker-prototype app: https://github.com/hypothesis/lms-file-picker-prototype/blob/master/src/server/index.js

**Request**

```
GET /authorize_lms_file_access
```

Authentication should use whatever mechanism is agreed for the LMS frontend to authenticate  requests to the backend, probably a query string param.

When this URL is opened in a popup, the backend should:

1. Construct a Canvas OAuth authorization request URL
2. Redirect the user to the Canvas OAuth page
3. Handle the redirect back to the LMS app when authorization completes. This will involve exchanging the auth code for an access token, saving the access token for the calling user
4. Closing the popup window after a successful auth flow.

**Response**

When the flow completes, whether successful or not, the server should render an HTML page which will call `window.close()` to close the popup window. Client-side code will detect this and continue.

The reason for closing the popup window rather than returning any response is that it makes the flow the same regardless of whether the user completes the auth flow successfully, or it fails, or the user interrupts it by closing the popup window.

**Comment by seanh** [comment]
[Canvas's docs on their authorization URL](https://canvas.instructure.com/doc/api/file.oauth.html#oauth2-flow-1)

**Comment by seanh** [comment]
To generate the authorization request URL we need:

1. The Canvas instance's base URL.

   Users enter this on the `/welcome` page when they register and we save it in the `application_instances` DB table. But there's little validation and no verification of this. There's no validation that what's entered is a URL. For example you can just enter `foo` and it'll be saved as `https://foo`. `The University of Foo` becomes `https://The University of Foo`. `https://` is prepended if no scheme is given. Trailing `/`'s are neither stripped nor appended. If the URL contains a path part that isn't stripped.

   There's also an LTI launch param `tool_consumer_instance_url` but it's an optional launch param in the spec and [not all LMS's support it](https://www.imsglobal.org/toolcompare.html).

   The app currently uses the string entered on the `/welcome` page.

2. Our `client_id` for this Canvas instance.

   This is the "Canvas Developer ID (optional)" that the user can enter on the `/welcome` page. It's the ID of a Canvas "developer key" that an admin has to manually create in Canvas. The key/secret part of the Canvas developer key (needed later as the OAuth 2 `client_secret`) is also entered into `/welcome` as "Canvas Developer Key (optional)". No validation or verification of these.

   These are optional and if none was entered when the application instance was registered at `/welcome` then we don't have Canvas Files support for that application instance. There's no way to add a developer ID and key later after registration (and fixing that is out of scope for this issue). 

3. Our `redirect_uri`.

   This is `https://lms.hypothes.is/canvas_oauth_callback`. I don't think we can change it because [user's enter it into Canvas when they create the developer key](https://web.hypothes.is/help/creating-canvas-developer-credentials-for-the-hypothesis-lms-app/) and Canvas will refuse any requests whose `redirect_uri` doesn't match that of the developer key (and we already have existing installs that've entered the existing `redirect_uri`).

   In fact the Canvas docs say the `redirect_uri` only has to have the same domain as that of the developer key. But I haven't tested that and it's probably safest to stick to the exact same URI.

4. `state` unique ID. We need to generate these and check them on redirect. Maybe we can use signed and timestamped JWTs rather than saving things to the DB.

5. `scope`, for supporting scoped developer keys. We should use this.

**Comment by seanh** [comment]
<https://github.com/hypothesis/lms/pull/571> implements a first proof-of-concept for this popup window endpoint, behind a feature flag. Once that's deployed there's still a few things to do before we can close this card:

- [ ] Store a `state` param in a cookie, and verify it when we receive the OAuth redirect from Canvas
- [ ] Handle variation and invalidness in the LMS URL's in our DB. These aren't validated at all on entry
- [ ] Handle application instances that have no developer key with an error response. The frontend code is gonna have to avoid showing the button in these cases anyway, but the auth endpoint should have an error response for it anyway
- [ ] Handle requests to the auth endpoint with a missing or invalid auth JWT, with an error page
- [ ] On receiving the redirect request send the access token request to the Canvas API
- [ ] Save the access and refresh token. Probably requires a new DB table
- [ ] Handle failures of the access token request
- [ ] Close the popup window after success

**Comment by seanh** [comment]
[Canvas's docs for their token endpoint](https://canvas.instructure.com/doc/api/file.oauth_endpoints.html#post-login-oauth2-token)

**Comment by seanh** [comment]
I'm gonna close this, I'm creating new cards for the remaining unchecked boxes
